{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# We will now start coding in python! We are going to load our counts data from featureCounts and make some files for DESeq2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "First we need to import all the python packages that we will use. We will do that in the following cell. **WARNING** you most likely won't have all these packages installed. If you get the error \"Cannot import package name ....\" You need to go back to your command line and install it on tscc. \n",
    "\n",
    "Remember how we installed things before? Try conda, then bioconda, then pip. Once you have installed it on your command line on tscc, you will need to restart your kernel to refresh those changes (Kernel - Restart)\n",
    "\n",
    "**NOTE - **since these are python specific packages, you will not be able to tell if they installed properly on your command line with \n",
    "\n",
    "    which packagename\n",
    "   \n",
    "Instead, you will first need to open python in your terminal with:\n",
    "\n",
    "    python\n",
    "    \n",
    "Then try to import it (this is the same thing you do in a Jupyter notebook)\n",
    "\n",
    "    import packagename\n",
    "    \n",
    "If you don't get any errors, then success! It is installed. To get out of python and back to your normal command line:\n",
    "\n",
    "    quit()\n",
    "    \n",
    "**Don't be afraid of error messages! Learn how to read them and solve them!!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "from matplotlib import pyplot as plt\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11, 4)\n",
      "(57820, 9)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chr</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Length</th>\n",
       "      <th>u2af1_control_k562_rep1.sorted.bam</th>\n",
       "      <th>u2af1_control_k562_rep2.sorted.bam</th>\n",
       "      <th>u2af1_shrna_k562_rep1.sorted.bam</th>\n",
       "      <th>u2af1_shrna_k562_rep2.sorted.bam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>chr1;chr1;chr1;chr1</td>\n",
       "      <td>11869;12595;12975;13221</td>\n",
       "      <td>12227;12721;13052;14412</td>\n",
       "      <td>+;+;+;+</td>\n",
       "      <td>1756</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...</td>\n",
       "      <td>14363;14970;15796;16607;16854;17233;17498;1760...</td>\n",
       "      <td>14829;15038;15947;16765;17055;17368;17504;1774...</td>\n",
       "      <td>-;-;-;-;-;-;-;-;-;-;-;-;-</td>\n",
       "      <td>2073</td>\n",
       "      <td>94</td>\n",
       "      <td>103</td>\n",
       "      <td>81</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>chr1;chr1;chr1</td>\n",
       "      <td>29554;30267;30976</td>\n",
       "      <td>30039;30667;31109</td>\n",
       "      <td>+;+;+</td>\n",
       "      <td>1021</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>chr1;chr1;chr1</td>\n",
       "      <td>34554;35245;35721</td>\n",
       "      <td>35174;35481;36081</td>\n",
       "      <td>-;-;-</td>\n",
       "      <td>1219</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>chr1;chr1</td>\n",
       "      <td>52473;54830</td>\n",
       "      <td>53312;54936</td>\n",
       "      <td>+;+</td>\n",
       "      <td>947</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                 Chr  \\\n",
       "Geneid                                                                 \n",
       "ENSG00000223972.4                                chr1;chr1;chr1;chr1   \n",
       "ENSG00000227232.4  chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;chr1;c...   \n",
       "ENSG00000243485.2                                     chr1;chr1;chr1   \n",
       "ENSG00000237613.2                                     chr1;chr1;chr1   \n",
       "ENSG00000268020.2                                          chr1;chr1   \n",
       "\n",
       "                                                               Start  \\\n",
       "Geneid                                                                 \n",
       "ENSG00000223972.4                            11869;12595;12975;13221   \n",
       "ENSG00000227232.4  14363;14970;15796;16607;16854;17233;17498;1760...   \n",
       "ENSG00000243485.2                                  29554;30267;30976   \n",
       "ENSG00000237613.2                                  34554;35245;35721   \n",
       "ENSG00000268020.2                                        52473;54830   \n",
       "\n",
       "                                                                 End  \\\n",
       "Geneid                                                                 \n",
       "ENSG00000223972.4                            12227;12721;13052;14412   \n",
       "ENSG00000227232.4  14829;15038;15947;16765;17055;17368;17504;1774...   \n",
       "ENSG00000243485.2                                  30039;30667;31109   \n",
       "ENSG00000237613.2                                  35174;35481;36081   \n",
       "ENSG00000268020.2                                        53312;54936   \n",
       "\n",
       "                                      Strand  Length  \\\n",
       "Geneid                                                 \n",
       "ENSG00000223972.4                    +;+;+;+    1756   \n",
       "ENSG00000227232.4  -;-;-;-;-;-;-;-;-;-;-;-;-    2073   \n",
       "ENSG00000243485.2                      +;+;+    1021   \n",
       "ENSG00000237613.2                      -;-;-    1219   \n",
       "ENSG00000268020.2                        +;+     947   \n",
       "\n",
       "                   u2af1_control_k562_rep1.sorted.bam  \\\n",
       "Geneid                                                  \n",
       "ENSG00000223972.4                                   0   \n",
       "ENSG00000227232.4                                  94   \n",
       "ENSG00000243485.2                                   0   \n",
       "ENSG00000237613.2                                   5   \n",
       "ENSG00000268020.2                                   0   \n",
       "\n",
       "                   u2af1_control_k562_rep2.sorted.bam  \\\n",
       "Geneid                                                  \n",
       "ENSG00000223972.4                                   1   \n",
       "ENSG00000227232.4                                 103   \n",
       "ENSG00000243485.2                                   1   \n",
       "ENSG00000237613.2                                   4   \n",
       "ENSG00000268020.2                                   0   \n",
       "\n",
       "                   u2af1_shrna_k562_rep1.sorted.bam  \\\n",
       "Geneid                                                \n",
       "ENSG00000223972.4                                 0   \n",
       "ENSG00000227232.4                                81   \n",
       "ENSG00000243485.2                                 0   \n",
       "ENSG00000237613.2                                 5   \n",
       "ENSG00000268020.2                                 0   \n",
       "\n",
       "                   u2af1_shrna_k562_rep2.sorted.bam  \n",
       "Geneid                                               \n",
       "ENSG00000223972.4                                 0  \n",
       "ENSG00000227232.4                                36  \n",
       "ENSG00000243485.2                                 1  \n",
       "ENSG00000237613.2                                 5  \n",
       "ENSG00000268020.2                                 0  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# This notebook is linked to TSCC, so we can load data directly by providing \n",
    "# the full path of the file location on tscc. Change the directory below to match your path\n",
    "data_directory = '/projects/ps-yeolab/biom262_2017/featurecounts/'\n",
    "save_directory = '/home/ucsd-train39/scratch/projects/u2af1_shrna/featurecounts/'\n",
    "\n",
    "# We are going to load the featurecounts file and assign it to the variable name: counts\n",
    "# change the \"featureCounts.txt\" to match the filename you used in your featureCounts script\n",
    "counts = pd.read_table(data_directory+\"all_counts.txt\",index_col=0,\n",
    "                    comment=\"#\")\n",
    "\n",
    "# Let's also load in the summary file so we can calculate the percent of reads\n",
    "# that aligned in our sample. \n",
    "\n",
    "summary = pd.read_table(data_directory+\"all_counts.txt.summary\",index_col=0,\n",
    "                       comment=\"#\")\n",
    "print summary.shape\n",
    "print counts.shape\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>u2af1_control_k562_rep1.sorted.bam</th>\n",
       "      <th>u2af1_control_k562_rep2.sorted.bam</th>\n",
       "      <th>u2af1_shrna_k562_rep1.sorted.bam</th>\n",
       "      <th>u2af1_shrna_k562_rep2.sorted.bam</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>1756</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>2073</td>\n",
       "      <td>94</td>\n",
       "      <td>103</td>\n",
       "      <td>81</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>1021</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>1219</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>947</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Length  u2af1_control_k562_rep1.sorted.bam  \\\n",
       "Geneid                                                          \n",
       "ENSG00000223972.4    1756                                   0   \n",
       "ENSG00000227232.4    2073                                  94   \n",
       "ENSG00000243485.2    1021                                   0   \n",
       "ENSG00000237613.2    1219                                   5   \n",
       "ENSG00000268020.2     947                                   0   \n",
       "\n",
       "                   u2af1_control_k562_rep2.sorted.bam  \\\n",
       "Geneid                                                  \n",
       "ENSG00000223972.4                                   1   \n",
       "ENSG00000227232.4                                 103   \n",
       "ENSG00000243485.2                                   1   \n",
       "ENSG00000237613.2                                   4   \n",
       "ENSG00000268020.2                                   0   \n",
       "\n",
       "                   u2af1_shrna_k562_rep1.sorted.bam  \\\n",
       "Geneid                                                \n",
       "ENSG00000223972.4                                 0   \n",
       "ENSG00000227232.4                                81   \n",
       "ENSG00000243485.2                                 0   \n",
       "ENSG00000237613.2                                 5   \n",
       "ENSG00000268020.2                                 0   \n",
       "\n",
       "                   u2af1_shrna_k562_rep2.sorted.bam  \n",
       "Geneid                                               \n",
       "ENSG00000223972.4                                 0  \n",
       "ENSG00000227232.4                                36  \n",
       "ENSG00000243485.2                                 1  \n",
       "ENSG00000237613.2                                 5  \n",
       "ENSG00000268020.2                                 0  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Let's drop unnecessary columns\n",
    "\n",
    "#Look at the column names in your dataframe above, we only want to keep Length and the four files\n",
    "#containing the count information. Make sure the names that you have after drop below match the \n",
    "#column names that you want to eliminate\n",
    "\n",
    "counts.drop(['Chr','Start','End','Strand'],axis=1, inplace=True)\n",
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Length</th>\n",
       "      <th>u2af1_control_k562_rep1</th>\n",
       "      <th>u2af1_control_k562_rep2</th>\n",
       "      <th>u2af1_shrna_k562_rep1</th>\n",
       "      <th>u2af1_shrna_k562_rep2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Geneid</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000223972.4</th>\n",
       "      <td>1756</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227232.4</th>\n",
       "      <td>2073</td>\n",
       "      <td>94</td>\n",
       "      <td>103</td>\n",
       "      <td>81</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000243485.2</th>\n",
       "      <td>1021</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000237613.2</th>\n",
       "      <td>1219</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000268020.2</th>\n",
       "      <td>947</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Length  u2af1_control_k562_rep1  u2af1_control_k562_rep2  \\\n",
       "Geneid                                                                        \n",
       "ENSG00000223972.4    1756                        0                        1   \n",
       "ENSG00000227232.4    2073                       94                      103   \n",
       "ENSG00000243485.2    1021                        0                        1   \n",
       "ENSG00000237613.2    1219                        5                        4   \n",
       "ENSG00000268020.2     947                        0                        0   \n",
       "\n",
       "                   u2af1_shrna_k562_rep1  u2af1_shrna_k562_rep2  \n",
       "Geneid                                                           \n",
       "ENSG00000223972.4                      0                      0  \n",
       "ENSG00000227232.4                     81                     36  \n",
       "ENSG00000243485.2                      0                      1  \n",
       "ENSG00000237613.2                      5                      5  \n",
       "ENSG00000268020.2                      0                      0  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Here we are going to change the name of the columns by providing a list of the \n",
    "#new column names we want to use. Order matters!\n",
    "\n",
    "counts = counts.rename(columns = {\"u2af1_control_k562_rep1.sorted.bam\":\"u2af1_control_k562_rep1\",\n",
    "                  \"u2af1_control_k562_rep2.sorted.bam\":\"u2af1_control_k562_rep2\",\n",
    "                  \"u2af1_shrna_k562_rep1.sorted.bam\":\"u2af1_shrna_k562_rep1\",\n",
    "                  \"u2af1_shrna_k562_rep2.sorted.bam\":\"u2af1_shrna_k562_rep2\"})\n",
    "counts.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#We will save this counts matrix to use as the input for differential expression. \n",
    "#This will be saved in the directory location that you defined above and will have the filename \n",
    "#counts_for_deseq2.csv\n",
    "\n",
    "counts.to_csv(save_directory+'counts_for_deseq2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>u2af1_control_k562_rep1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_control_k562_rep2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_shrna_k562_rep1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_shrna_k562_rep2</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [u2af1_control_k562_rep1, u2af1_control_k562_rep2, u2af1_shrna_k562_rep1, u2af1_shrna_k562_rep2]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#We also need to make a condition matrix for deseq2. The row names need to match the column names\n",
    "#from the count matrix and we will assign a condition to each row\n",
    "\n",
    "conditions = pd.DataFrame(index = counts.columns[1:])\n",
    "conditions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>condition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>u2af1_control_k562_rep1</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_control_k562_rep2</th>\n",
       "      <td>control</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_shrna_k562_rep1</th>\n",
       "      <td>knockdown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>u2af1_shrna_k562_rep2</th>\n",
       "      <td>knockdown</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         condition\n",
       "u2af1_control_k562_rep1    control\n",
       "u2af1_control_k562_rep2    control\n",
       "u2af1_shrna_k562_rep1    knockdown\n",
       "u2af1_shrna_k562_rep2    knockdown"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Make sure the features listed below match the order of the rows that are listed in the \n",
    "#conditions dataframe that is printed above. If necessary, change the order to match your\n",
    "#setup. Check that the conditions match on the dataframe created below\n",
    "\n",
    "features= ['control','control','knockdown','knockdown']\n",
    "conditions['condition'] = features\n",
    "conditions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#We will save this counts matrix to use as the input for differential expression. \n",
    "#This will be saved in the directory location that you defined above and will have the filename \n",
    "#conditions_for_deseq2.csv\n",
    "\n",
    "conditions.to_csv(save_directory+'conditions_for_deseq2.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
